8333446: Add tests for hierarchical container support

jerboaa opened the pull request:
Please review this PR which adds test support for systemd slices so that bugs like [JDK-8217338](https://bugs.openjdk.org/browse/JDK-8217338) can be verified. The added test, `SystemdMemoryAwarenessTest` currently passes on cgroups v1 and fails on cgroups v2 due to the way how [JDK-8217338](https://bugs.openjdk.org/browse/JDK-8217338) was implemented when JDK 13 was a thing. Therefore immediately problem-listed. It should get unlisted once [JDK-8322420](https://bugs.openjdk.org/browse/JDK-8322420) merges.

I'm adding those tests in order to not regress another time.

Testing:
- [x] Container tests on Linux x86_64 cgroups v2 and Linux x86_64 cgroups v1.
- [x] New systemd test on cg v1 (passes). Fails on cg v2 (due to  JDK-8322420)
- [ ] GHA

jerboaa commented:
GHA failure of macos-x64 seems infra related and not related to this patch.

jerboaa commented:
Anyone willing to review this?

jerboaa commented:
Gentle ping.

jankratochvil commented on test/hotspot/jtreg/ProblemList.txt:
```diff
@@ -116,6 +116,7 @@ applications/jcstress/copy.java 8229852 linux-all
 containers/docker/TestJcmd.java 8278102 linux-all
 containers/docker/TestMemoryAwareness.java 8303470 linux-all
 containers/docker/TestJFREvents.java 8327723 linux-x64
+containers/systemd/SystemdMemoryAwarenessTest.java 8322420 linux-all
```
This line should be removed as long as it gets applied after [17198](https://github.com/openjdk/jdk/pull/17198).

jerboaa replied:
Sure. We need to see which one goes in first and I'll adjust accordingly.

jankratochvil commented:
[test.patch.txt](https://github.com/user-attachments/files/16171122/test.patch.txt)
* `CPUQuota` (changed it to `AllowedCPUs`) does not work for me - it properly distributes the load but JDK still sees all available CPU cores (4 of my VM).
* the change 2 -> 1 cores: // We could check 2 cores ("0-1") but then it would fail on single-core nodes / virtual machines.

jerboaa commented:
> [test.patch.txt](https://github.com/user-attachments/files/16171122/test.patch.txt)
> 
>     * `CPUQuota` (changed it to `AllowedCPUs`) does not work for me - it properly distributes the load but JDK still sees all available CPU cores (4 of my VM).

Could you elaborate on that? What does not work? It's relying on the JVM properly detecting the set limit. `CPUQuota` sets the values in `cpu.max` on unified hierarchy for the `cpu` controller. See the [systemd doc](https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html). It's available since systemd 213. RHEL 7 has 219 which should be good enough. `AllowedCPUs` on the other hand uses the `cpuset` controller, which is a different thing. For the purpose of this test, we should use `CPUQuota`.
 
>     * the change 2 -> 1 cores: // We could check 2 cores ("0-1") but then it would fail on single-core nodes / virtual machines.

Yeah, we have a chicken/egg problem there. It seemed assuming 2 cores is reasonable. We could query the number of not restricted CPUs (of the physical system) using the WB API and take the minimum of the two. Let me work on that.

jankratochvil commented:
> > ```
> > * `CPUQuota` (changed it to `AllowedCPUs`) does not work for me - it properly distributes the load but JDK still sees all available CPU cores (4 of my VM).
> > ```
> 
> Could you elaborate on that? What does not work?

In the log there is (`/proc/cpuinfo` has 4 entries on this system):
```
[0.139s][trace][os,container] OSContainer::active_processor_count: 4 
```
and therefore it fails with:
```    
java.lang.RuntimeException: 'OSContainer::active_processor_count: 2' missing from stdout/stderr
        at jdk.test.lib.process.OutputAnalyzer.shouldContain(OutputAnalyzer.java:252)
        at SystemdMemoryAwarenessTest.testHelloSystemd(SystemdMemoryAwarenessTest.java:58)
        at SystemdMemoryAwarenessTest.main(SystemdMemoryAwarenessTest.java:43)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
        at java.base/java.lang.Thread.run(Thread.java:1575)
```
It is on Fedora 40 x86_64 (`systemd-255.8-1.fc40.x86_64`).

jerboaa commented:
> > > ```
> > > * `CPUQuota` (changed it to `AllowedCPUs`) does not work for me - it properly distributes the load but JDK still sees all available CPU cores (4 of my VM).
> > > ```
> > 
> > 
> > Could you elaborate on that? What does not work?
> 
> In the log there is (`/proc/cpuinfo` has 4 entries on this system):
> 
> ```
> [0.139s][trace][os,container] OSContainer::active_processor_count: 4 
> ```
> 
> and therefore it fails with:
> 
> ```
> java.lang.RuntimeException: 'OSContainer::active_processor_count: 2' missing from stdout/stderr
>         at jdk.test.lib.process.OutputAnalyzer.shouldContain(OutputAnalyzer.java:252)
>         at SystemdMemoryAwarenessTest.testHelloSystemd(SystemdMemoryAwarenessTest.java:58)
>         at SystemdMemoryAwarenessTest.main(SystemdMemoryAwarenessTest.java:43)
>         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
>         at java.base/java.lang.reflect.Method.invoke(Method.java:580)
>         at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
>         at java.base/java.lang.Thread.run(Thread.java:1575)
> ```
> 
> It is on Fedora 40 x86_64 (`systemd-255.8-1.fc40.x86_64`).

Well yes, because the limit isn't properly detected (needs a JVM change that does that; imo https://github.com/openjdk/jdk/pull/17198).

jerboaa commented:
Example test run on cgv1 with a fixed JVM: https://cr.openjdk.org/~sgehwolf/webrevs/jdk-8333446-systemd-slice-tests/cgv1/SystemdMemoryAwarenessTest.jtr
Example test run on cgv2 with a fixed JVM: https://cr.openjdk.org/~sgehwolf/webrevs/jdk-8333446-systemd-slice-tests/cgv2/SystemdMemoryAwarenessTest.jtr

jerboaa commented:
Please keep it open, bot.

jerboaa commented:
GHA failure on maxos-aarch64 is unrelated: [runtime/cds/DeterministicDump](https://github.com/jerboaa/jdk/actions/runs/10476366525#user-content-runtime_cds_deterministicdump)

jerboaa commented:
@dholmes-ora @iklam @MBaesken Could somebody of you help review this, please? Increases test coverage in the container detection area. Thanks!

zzambers commented:
If I am not mistaken, new test requires, that testsuite is ran as superuser (root). (Because it writes `/etc/systemd/system`, runs certain systemd commands). Should test be skipped for non-root?

MBaesken commented:
I added the PR to our internal build/test queue .

jerboaa commented:
> If I am not mistaken, new test requires, that testsuite is ran as superuser (root). (Because it writes `/etc/systemd/system`, runs certain systemd commands). Should test be skipped for non-root?

Thanks! I can add that. FWIW, container tests are in a similar situation (applying cpu/memory limits is not allowed in rootless on cg v1) and they don't check for it.

jerboaa commented:
> I added the PR to our internal build/test queue .

Thanks, Matthias!

zzambers commented:
> > If I am not mistaken, new test requires, that testsuite is ran as superuser (root). (Because it writes `/etc/systemd/system`, runs certain systemd commands). Should test be skipped for non-root?
> 
> Thanks! I can add that. FWIW, container tests are in a similar situation (applying cpu/memory limits is not allowed in rootless on cg v1) and they don't check for it.

I think, it would be nice to skip on non-root, as tests are often ran as non-root. Thanks.

MBaesken commented on test/hotspot/jtreg/containers/systemd/SystemdMemoryAwarenessTest.java:
```diff
@@ -0,0 +1,87 @@
+    private static void testHelloSystemd() throws Exception {
+        SystemdRunOptions opts = SystemdTestUtils.newOpts("HelloSystemd");
+        // 1 GB memory
+        opts.memoryLimit("1000M");
```
Just wondering - is 1G here possible (the comment states 1 GB / 1024M) ?

jerboaa replied:
I probably shall fix the comment or change it to `1024M`. Either way it has to match the assertion where we look for `1048576000` bytes in the output.

jerboaa replied:
This should be fixed now.

jerboaa commented:
Noting here that I'll amend those tests to also cover nested hierarchical limits where the lower limit is higher up the hierarchy.

MBaesken commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2497,6 +2504,12 @@ WB_ENTRY(jint, WB_ValidateCgroup(JNIEnv* env,
   return ret;
 WB_END
 
+// Physical cpus of the host machine (including containers), Linux only.
```
Isn't the comment a bit misleading ?  From what I see , ` os::Linux::active_processor_count()` can use various mechanisms to get number of processor info, if it uses https://linux.die.net/man/2/sched_getaffinity it gives the 'set of CPUs on which it is eligible to run.'   That might be different from what the host has.

jerboaa replied:
Yes. See #20768 for an attempt to unify it. I'll change the comment with the update that I have for nested hierarchies. Thanks!

jerboaa replied:
I've changed the comment.

MBaesken commented:
Looking through the coding it looks more or less okay to me; but if you really need to run it under user 'root'  I think we will not have so much use for this in our test environments  because we use other test users.
Not saying that this is a very bad thing,  maybe it is just the way it is, that 'root' is needed ?

jerboaa commented:
> Not saying that this is a very bad thing, maybe it is just the way it is, that 'root' is needed ?

I'll do some more research whether or not that is a hard requirement. Thanks for the comments so far.

jerboaa commented:
> > Not saying that this is a very bad thing, maybe it is just the way it is, that 'root' is needed ?
> 
> I'll do some more research whether or not that is a hard requirement. Thanks for the comments so far.

It turns out it works on cgroups v2 as non-root. I shall amend the test so that it at least runs OK on non-root and cgroups v2.

jerboaa commented:
@MBaesken Any more thoughts on this?

jerboaa commented:
Thank you for the review, @MBaesken!

@zzambers OK for you as well?

zzambers commented:
I have done some testing on RHELs (build with changes from this PR + other 2 cgroup PRs applied):
**RHEL-8** (cgroup1/non-root)
- test was skipped correctly

**RHEL-9** (cgroup2/non-root)
- I saw failure of `active_processor_count` check.
- after investigation, I have found, that `cpu` cgroup controller is not delegated to `user@1000.service` (and children) on rhel-9 (unlike in e.g. fedora) it only had `memory pids` (btw. available controllers at given "level" are listed in `cgroup.controllers` file in cgroups v2)
- when I modified `user@.service` to also delegate cpu controller, test passed

Apart from issue with check for `active_processor_count` on RHEL-9/non-root, it looks good. However I don't know how to easily fix issue with `active_processor_count`. Maybe check could be skipped for non-root. (Work-around is to modify system configuration.)

jerboaa commented:
@zzambers Thanks for taking a look.

> I have done some testing on RHELs (build with changes from this PR + other 2 container PRs applied): **RHEL-8** (cgroup1/non-root)
> 
>     * test was skipped correctly
> 
> 
> **RHEL-9** (cgroup2/non-root)
> 
>     * I saw failure of `active_processor_count` check.
> 
>     * after investigation, I have found, that `cpu` cgroup controller is not delegated to `user@1000.service` (and children) on rhel-9 (unlike in e.g. fedora) it only had `memory pids` (btw. available controllers at given "level" are listed in `cgroup.controllers` file in cgroups v2)
> 
>     * when I modified `user@.service` to also delegate cpu controller, test passed

Could it be that the setup you've done to employ delegation is similar to this one?
https://github.com/jerboaa/openjdk-cgroupv2-setup/blob/97690683af17b303276ea473fe44b3dde7ead327/config_cgroupv2.yml#L24-L32
 
> Apart from issue with check for `active_processor_count` on RHEL-9/non-root, it looks good. However I don't know how to easily fix issue with `active_processor_count` check. Maybe check could be skipped for non-root. (Work-around is to modify system configuration.)

Do existing podman container tests pass on that system? It seems fair to assume that that's the baseline config for container tests in general: systemd ones or podman/docker. I know that on cg v2 not all container tests pass out-of-the-box. In particular certain CPU awareness tests. Keeping that basic idea in terms of required config for those tests consistent with other container tests seem adequate to me.

zzambers commented:
> @zzambers Thanks for taking a look.
> 
> > I have done some testing on RHELs (build with changes from this PR + other 2 container PRs applied): **RHEL-8** (cgroup1/non-root)
> > ```
> > * test was skipped correctly
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **RHEL-9** (cgroup2/non-root)
> > ```
> > * I saw failure of `active_processor_count` check.
> > 
> > * after investigation, I have found, that `cpu` cgroup controller is not delegated to `user@1000.service` (and children) on rhel-9 (unlike in e.g. fedora) it only had `memory pids` (btw. available controllers at given "level" are listed in `cgroup.controllers` file in cgroups v2)
> > 
> > * when I modified `user@.service` to also delegate cpu controller, test passed
> > ```
> 
> Could it be that the setup you've done to employ delegation is similar to this one? https://github.com/jerboaa/openjdk-cgroupv2-setup/blob/97690683af17b303276ea473fe44b3dde7ead327/config_cgroupv2.yml#L24-L32

I have just added `cpu` to Delegate list of `user@.service`, looks similar, to what is done there. I see use of `Delegate=yes` in your link, that probably delegates all.
Thanks for this link. 

> 
> > Apart from issue with check for `active_processor_count` on RHEL-9/non-root, it looks good. However I don't know how to easily fix issue with `active_processor_count` check. Maybe check could be skipped for non-root. (Work-around is to modify system configuration.)
> 
> Do existing podman container tests pass on that system? It seems fair to assume that that's the baseline config for container tests in general: systemd ones or podman/docker. I know that on cg v2 not all container tests pass out-of-the-box. In particular certain CPU awareness tests. Keeping that basic idea in terms of required config for those tests consistent with other container tests seem adequate to me.

You are right. I have ran container tests in my VM and indeed faced issue with missing cpuset controller.
(yesterday I forgot to set required properties, so most of them got skipped)

Interesting that we have not faced this issue in our testing (container tests are passing). However that is probably because we run containers tests in different way (we don't use VMs for it, but rather run them in beaker). I would need to investigate.

Anyway good to know, there can be this issue with cgroup controllers.

jerboaa commented:
> I have just added `cpu` to Delegate list of `user@.service`, looks similar, to what is done there. I see use of `Delegate=yes` in your link, that probably delegates all.
> Thanks for this link.

FWIW, I was able to reproduce what you said on RHEL 9 when run as user. The mentioned config fixes it. I'll add a hint when the assertion fails.

jerboaa commented:
@MBaesken @zzambers I've updated this patch which should cover the config issue as pointed out by @zzambers. The latest version throws `SkippedException` with a hint should the match fail. Done in https://github.com/openjdk/jdk/pull/19530/commits/0e52e004e9766301294743aa42c8306d7a25a34f and added this info to the bug as well.

Good to go now?

jerboaa commented:
I've merged master and removed the test from the problem list since the relevant bug got fixed.

@MBaesken Said changes require a re-review before this patch would be ready for integration. Could you please take another look? Thank you!

jerboaa commented:
Thanks all for the re-reviews!

jerboaa commented:
/integrate
